### DISCUSSION TITLE
Tutorial won't compile:   Meet The Composable Architecture - Testing your Feature, Testing Effects, Step 7 has an error

### DISCUSSION THREAD

**Discussion by awolf-foreflight** [pointfreeco/swift-composable-architecture#2271] [General]
### Description

The tutorial has an error on line 14.   This won't compile for me with Xcode 14.3.1 despite the store defined on line 7 being an instance of TestStore.

You can see the autocompleted doesn't provide the expected receive method.  When I paste it in the code from the tutorial the test won't compile

<img width="851" alt="image" src="https://github.com/pointfreeco/swift-composable-architecture/assets/131712912/19155504-d1bd-4b7b-89da-a81f974d65af">

# This is a screen grab of the tutorial:
<img width="1190" alt="image" src="https://github.com/pointfreeco/swift-composable-architecture/assets/131712912/2fd1b9b0-702e-4d72-b6e1-d5bb945e639b">

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

Copying and pasting code from a tutorial should compile

### Actual behavior

Multiple compiler errors

This screen grab shows if you simply instantiate a test store with `TestStore()` then autocomplete works and the test compiles

<img width="1390" alt="image" src="https://github.com/pointfreeco/swift-composable-architecture/assets/131712912/ac0a8d42-b837-4212-8b29-c93e6925ea0e">

But using the `store` constant doesn't work. 🤷‍♂️

The root cause appears to be that the tutorial doesn't tell us to add the following line to the `CounterFeature.swift` file.
```
extension CounterFeature.Action: Equatable {}

```
this fixes the problem

I got lucky and got an error about Action needing to be equatable and remembered the trick the tutorial required to make `State` equatable.

But this is a frustrating experience and difficult to resolve.

### Steps to reproduce

Copy paste tutorial example code, try to compile, observe compiler error

### The Composable Architecture version information

0.55.0

### Destination operating system

ios 16.3

### Xcode version information

14.3.1

### Swift Compiler version information

```shell
awolf@Lins-iPhone-5s ios % xcrun swiftc --version
swift-driver version: 1.75.2 Apple Swift version 5.8.1 (swiftlang-5.8.0.124.5 clang-1403.0.22.11.100)
Target: arm64-apple-macosx13.0
```

**Comment by mbrandonw** [comment]
Hi @awolf-foreflight, a few steps before the step you highlight it was mentioned that the actions need to be made `Equatable`:

https://github.com/pointfreeco/swift-composable-architecture/blob/af16494b9e6ccd306605aa328f5deea3d75cd33a/Sources/ComposableArchitecture/Documentation.docc/Tutorials/MeetTheComposableArchitecture/01-Essentials/03-TestingYourFeatures/01-03-TestingYourFeature.tutorial#L157-L158

Maybe this could be improved, but it's not really an issue with the library so I am going to convert this to a discussion where we can further discuss ways to improve this.

**Reply by awolf-foreflight** [reply]
Hi @mbrandonw - You're right, there is a "Note" that mentions this.    Thank you for closing this out and making it a discussion. 

My bad for not reading more carefully.

I also feel the tutorial is partially at fault too because:
* Every compiler error until this point was listed in the code as comments
* I think the tutorial even stated "it won't compile yet" or said to build and observe the error
* Skimming all previous notes did not result in a build error.

Also consider that I was typing the code in, not copy/pasting in this one case case. The lack of autocomplete was baffling.  And there was no compiler error to help me.

I eventually just copy/pasted and it was only then I got a compiler error.   I was over an hour into the tutorial at the point.

I feel it is not entirely unreasonable at this point to assume all of the code I need to compile is provided because that had been the case for dozens and dozens of steps prior.

I acknowledge the alternate view that many folks have made it through just fine as it is.   I provided the info on how I actually experienced it to illustrate that this may have a negative impact to anyone who encounters it as I did.

-----------
In closing - the way scrolling adds in the new code and highlights it is really, really fantastic.  I love the way you did this.   I think an extra step to jump over and add that line of code into the Feature would be a nice enhancement and be consistent with all of the other steps and notes in the tutorial

Either way, thanks for the tutorial.   It is great and very, very helpful.

**Reply by mbrandonw** [reply]
Hi @awolf-foreflight, just to clarify, this is what step 5 looks like in the documentation:

<img width="809" alt="image" src="https://github.com/pointfreeco/swift-composable-architecture/assets/135203/e75356e7-218b-4855-a318-d3d101130d70">


So not only does the "Note" detail that actions need to be equatable, but the code also shows this with the `extension` snippet.

I understand how frustrating it can be to get compiler errors while walking through the tutorial, but it is not immediately clear to me how we can improve the situation. We are definitely open to suggestions.

Also, FWIW, all of our tutorials are built on top of Apple's DocC tool, of which there are many WWDC videos about. And the markdown files are all in [this](https://github.com/pointfreeco/swift-composable-architecture/tree/main/Sources/ComposableArchitecture/Documentation.docc) directory. If you ever find typos or errors it will always be welcome to receive a pull request.

**Reply by awolf-foreflight** [reply]
Hi @mbrandonw - 

I see the exact same.  I was so certain that it wasn't there previously I checked the gitHub history.   And this is my error. 🤦‍♂️. According to gitHub the main file and the two code files (004 & 005) have not been changed in two months.  

I'm very sorry about that, I went through this multiple times - yet here it is.  It is no excuse, but the best I can figure is I was looking for it to jump to the `CounterFeature.swift` file to add the conformance.  

Thank you for pointing out my error - this is exactly what I was hoping for.
<img width="1419" alt="image" src="https://github.com/pointfreeco/swift-composable-architecture/assets/131712912/91be6ce3-42fc-4866-8d0e-f8dcadf61c6c">

**Comment by awolf-foreflight** [comment]
Hi again @mbrandonw  - I almost forgot.  The reason I was back on GitHub looking at my issue was because of the following.   Another example in the screenshot where the code suddenly doesn't compile.   I don't understand why and now I'm stuck.

I'm trying to prepare for a meeting tomorrow and now I have to switch gears and go figure out something I probably already know - I'm just not seeing the issue.

Oh, I just found it.  [Rubber Duckie](https://en.wikipedia.org/wiki/Rubber_duck_debugging) FTW!  🤓

```
    enum Action: Equatable {
        
        case cancelButtonTapped
        case delegate(Delegate)
        case saveButtonTapped
        case setName(String)
        enum Delegate: Equatable { <<<<< the tutorial is missing the "Equatable" conformance
            case cancel
            case saveContact(Contact)
        }
    }
```


Despite the compiler correctly describing the problem this time **and** having written this same bug a few times - I still didn't see it until I was writing this post.

This aligns well with my previous post.  Switching gears from "wrestling with TCA and trying to understand it" to debugging a random compiler error is not ideal.  Worse it completely blocks further progress on the tutorial.



<img width="2226" alt="image" src="https://github.com/pointfreeco/swift-composable-architecture/assets/131712912/4d1eebf9-8b61-41af-9ff3-cce337fe02b6">

**Reply by mbrandonw** [reply]
This is definitely an accidental omission in the tutorial and I have just opened a PR https://github.com/pointfreeco/swift-composable-architecture/pull/2280 to address it.

**Reply by awolf-foreflight** [reply]
Hi @mbrandonw - I didn't realize the association of code with each step (and the diffing) was part of DoC.   That's remarkable.

If I think I see an issue in the future I will first quadruple check myself.  If I still think there is a problem I'll submit a PR for your review along with my feedback question.   

I reviewed your PR, checked out the branch, and build the Docs locally - I believe I see how it works.  But given the above I'll be questioning myself on everything for the rest of the day....

Thanks again for your help (and your time).
